lib/Runtime/Library/BoundFunction.cpp

```diff
@@ -136,16 +136,24 @@ namespace Js
         Var newVarInstance = nullptr;
         if (callInfo.Flags & CallFlags_New)
         {
-          if (JavascriptProxy::Is(targetFunction))
-          {
-            JavascriptProxy* proxy = JavascriptProxy::FromVar(targetFunction);
-            Arguments proxyArgs(CallInfo(CallFlags_New, 1), &targetFunction);
-            args.Values[0] = newVarInstance = proxy->ConstructorTrap(proxyArgs, scriptContext, 0);
-          }
-          else
-          {
-            args.Values[0] = newVarInstance = JavascriptOperators::NewScObjectNoCtor(targetFunction, scriptContext);
-          }
+            if (args.HasNewTarget())
+            {
+                // target has an overriden new target make a new object from the newTarget
+                RecyclableObject* newTarget = RecyclableObject::UnsafeFromVar(args.Values[args.Info.Count]);
```

Comment:
What code is responsible for guaranteeing that newTarget is actually a RecyclableObject? Since it's not nearby, we can avoid a potential security issue by calling `FromVar` instead of `UnsafeFromVar`, and I think it's worth the extra runtime cost. Otherwise, if we have a strict guarantee that newTarget is always a RecyclableObject, then we should update GetNewTarget so that it returns a RecyclableObject and doesn't require calling code to do an unsafe cast.

Reply:
Hmm - agreed I haven't checked this thoroughly enough - rather than a fail fast, if it's ok I'll duplicate what Reflect.construct does and throw a TypeError if it's not a constructor?

Reply:
Seems reasonable to me.

---
In reply to: [188105251](https://github.com/Microsoft/ChakraCore/pull/5147#discussion_r188105251) [](ancestors = 188105251)

Reply:
I think that going with a FailFast is reasonable here; my understanding of `newTarget` is that it is supposed to be the (most-derived / final) constructor function, and so it shouldn't be possible to get a non-constructible function here. If we're breaking that assumption, I'd at least want to see a debug assertion, and for safety I think a failfast is appropriate.

Reply:
@MSLaguana `new.target` is either the exact function called with `new` (always constructible by definition) or else the value specified for newTarget in a `Reflect.construct` call.  In that latter case I think it could be non-constructible?

Reply:
From my experimentation, if you pass something that isn't a function, or is a non-constructible function (e.g. an arrow function) then `Reflect.construct` throws an error. That's why I think it should always be a constructible function, and I'd like a fairly noisy error if it's not the case 😄

Reply:
Indeed there's a runtime error in Reflect.construct if you pass anything that's not a constructor as newTarget:
https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Library/JavascriptReflect.cpp#L383-L403